libia2/include/ia2.h

```diff
@@ -226,28 +228,125 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+  /* Returns the top of the stack, not the base address of the allocation. */  \
+  char *allocate_stack(int i) {                                                \
     char *stack = (char *)mmap(NULL, STACK_SIZE, PROT_READ | PROT_WRITE,       \
                                MAP_PRIVATE | MAP_ANON, -1, 0);                 \
     if (stack == MAP_FAILED) {                                                 \
       printf("Failed to allocate stack %d (%s)\n", i, errno_s);                \
       exit(-1);                                                                \
     }                                                                          \
-    return stack;                                                              \
+    if (i != 0) {                                                              \
+      int res = pkey_mprotect(stack, STACK_SIZE, PROT_READ | PROT_WRITE, i);   \
+      if (res == -1) {                                                         \
+        printf("Failed to mprotect stack %d (%s)\n", i, errno_s);              \
+        exit(-1);                                                              \
+      }                                                                        \
+    }                                                                          \
+    /* Each stack frame start + 8 is initially 16-byte aligned. */             \
+    return stack + STACK_SIZE - 8;                                             \
+  }                                                                            \
+  void init_tls_15(void);                                                      \
+  void init_tls_14(void);                                                      \
+  void init_tls_13(void);                                                      \
+  void init_tls_12(void);                                                      \
+  void init_tls_11(void);                                                      \
+  void init_tls_10(void);                                                      \
+  void init_tls_9(void);                                                       \
+  void init_tls_8(void);                                                       \
+  void init_tls_7(void);                                                       \
+  void init_tls_6(void);                                                       \
+  void init_tls_5(void);                                                       \
+  void init_tls_4(void);                                                       \
+  void init_tls_3(void);                                                       \
+  void init_tls_2(void);                                                       \
+  void init_tls_1(void);                                                       \
+  /* Ensure that TLS is protected in a new thread. */                          \
+  void protect_tls(void) {                                                     \
```

Comment:
This is a strong, public symbol. Won't it conflict between shared libs? This could use better documentation of why you did the goto thing here, why not just call `protect_tls_for_compartment(n)`?

Reply:
It will not conflict between shared libs because this definition is in the `INIT_RUNTIME_COMMON` macro; `INIT_RUNTIME` should only appear in the main compartment, so there is only one definition of the function.

The way we have our code laid out and split between libia2 itself and the main compartment is really not ideal. `ia2.h` contains:
- IA2 interface declarations, like a normal header (e.g. for internal use by libia2 to refer to symbols present in the main compartment or in libia2 itself)
- definitions of macros that have the side-effect of defining per-compartment symbols such as `init_pkey_##n##_ctor`
- definitions of macros that have the side-effect of defining per-program symbols built into the main compartment, such as `protect_tls` and `ia2_stackptr_0`

It gets confusing which definitions are where for what reasons.

I'll document the `goto` hack. We do an equivalent thing with `switch` in `init_stacks`, and I couldn't decide which is cleaner... I have both for variety and hoped to get some feedback on which is preferable. The reason either hack is necessary is that we want to unroll the pseudo-loop of `for(int i=0; i<n; i++) { macro(i) }` for a compile-time integer `n`, in order to get straight-line code that only does the required work for the number of compartments that the program actually uses. For example, from `two_keys_minimal`:
```asm

00000000000028f0 <init_stacks>:
    28f0:	41 54                	push   %r12
    28f2:	bf 02 00 00 00       	mov    $0x2,%edi
    28f7:	53                   	push   %rbx
    28f8:	48 83 ec 08          	sub    $0x8,%rsp
    28fc:	e8 af fd ff ff       	call   26b0 <allocate_stack>
    2901:	49 89 c2             	mov    %rax,%r10
    2904:	31 c0                	xor    %eax,%eax
    2906:	89 c3                	mov    %eax,%ebx
    2908:	89 c1                	mov    %eax,%ecx
    290a:	89 c2                	mov    %eax,%edx
    290c:	41 89 c4             	mov    %eax,%r12d
    290f:	b8 cc ff ff ff       	mov    $0xffffffcc,%eax
    2914:	4c 8b 1d dd 26 00 00 	mov    0x26dd(%rip),%r11        # 4ff8 <ia2_stackptr_2@Base>
    291b:	64 49 83 3b 00       	cmpq   $0x0,%fs:(%r11)
    2920:	74 0e                	je     2930 <init_stacks+0x40>
    2922:	48 c7 c7 02 00 00 00 	mov    $0x2,%rdi
    2929:	e8 92 ff ff ff       	call   28c0 <ia2_reinit_stack_err>
    292e:	0f 0b                	ud2
    2930:	64 4d 89 13          	mov    %r10,%fs:(%r11)
    2934:	44 89 e0             	mov    %r12d,%eax
    2937:	bf 01 00 00 00       	mov    $0x1,%edi
    293c:	e8 6f fd ff ff       	call   26b0 <allocate_stack>
    2941:	49 89 c2             	mov    %rax,%r10
    2944:	31 c0                	xor    %eax,%eax
    2946:	89 c3                	mov    %eax,%ebx
    2948:	89 c1                	mov    %eax,%ecx
    294a:	89 c2                	mov    %eax,%edx
    294c:	41 89 c4             	mov    %eax,%r12d
    294f:	b8 f0 ff ff ff       	mov    $0xfffffff0,%eax
    2954:	4c 8b 1d 75 26 00 00 	mov    0x2675(%rip),%r11        # 4fd0 <ia2_stackptr_1@@Base+0x4fd0>
    295b:	64 49 83 3b 00       	cmpq   $0x0,%fs:(%r11)
    2960:	74 0e                	je     2970 <init_stacks+0x80>
    2962:	48 c7 c7 01 00 00 00 	mov    $0x1,%rdi
    2969:	e8 52 ff ff ff       	call   28c0 <ia2_reinit_stack_err>
    296e:	0f 0b                	ud2
    2970:	64 4d 89 13          	mov    %r10,%fs:(%r11)
    2974:	44 89 e0             	mov    %r12d,%eax
    2977:	31 ff                	xor    %edi,%edi
    2979:	e8 32 fd ff ff       	call   26b0 <allocate_stack>
    297e:	48 89 c3             	mov    %rax,%rbx
    2981:	64 48 8b 04 25 00 00 00 00 	mov    %fs:0x0,%rax
    298a:	48 03 05 2f 26 00 00 	add    0x262f(%rip),%rax        # 4fc0 <ia2_stackptr_0@@Base+0x4fb8>
    2991:	48 89 18             	mov    %rbx,(%rax)
    2994:	48 83 c4 08          	add    $0x8,%rsp
    2998:	5b                   	pop    %rbx
    2999:	41 5c                	pop    %r12
    299b:	c3                   	ret
```

Reply:
We can, at least, namespace this as `ia2_protect_tls`.

Reply:
I documented the static dispatch hack. I'd still like to settle on one of `goto label##n`/`switch(literal)`, and I'll namespace the symbol.